src/stories/Forms/Select/Select.tsx

```diff
@@ -0,0 +1,489 @@
+    if (!React.isValidElement(child)) return false;
+    return child.type === Option;
+  });
+};
+
+const getSelectContentAsString = (node: string | React.ReactNode): string | null => {
+  if (!node) return null;
+  if (typeof node === 'string') return node;
+  return React.Children.map(node, (child) => {
+    if (typeof child === 'string') return child;
+    if (!React.isValidElement(child)) return;
+    const innerChild = child.props.children;
+    if (typeof innerChild === 'string' || typeof innerChild === 'number') return innerChild;
+  }).join(' ');
+};
+
+const getChildrenByQuery = (children: SelectProps['children'], query: string) => {
+  return React.Children.toArray(children).filter((child: React.ReactElement<OptionProps>) => {
+    if (!React.isValidElement(child)) return true;
+
+    const searchContent = [];
+    if (typeof child.props.children === 'string') {
+      searchContent.push(child.props.children);
+    }
+    if (child.props.value) {
+      searchContent.push(child.props.value);
+    }
+    const innerSearchContent = React.Children.map(child.props.children, (subchild) => {
+      if (typeof subchild === 'string') return subchild;
+      if (!React.isValidElement(subchild)) return;
+      if (typeof subchild.props.children !== 'string') return;
+      return subchild.props.children;
+    });
+    searchContent.push(...innerSearchContent);
+
+    const searchContentString = searchContent.join(' ').toLowerCase();
+    return searchContentString.indexOf(query.toLowerCase()) !== -1;
+  });
+};
+
+export interface SelectProps extends GlobalInputProps {
+  /**
+   * Option components to be rendered.
+   */
+  children: React.ReactNode;
+  showItems?: number;
+}
+
+/**
+ * Select component
+ */
+export const Select = React.forwardRef((props: SelectProps, ref: React.RefObject<HTMLInputElement>) => {
+  const {
+    ariaDescribedBy,
+    ariaErrorMessage,
+    children,
+    className,
+    hasError,
+    id,
+    isRequired,
+    name,
+    onBlur,
+    onChange,
+    showItems = 6,
+    value,
+    ...additionalPhoneInputProps
+  } = props;
+
+  if (!validateChildren(children)) {
+    throw new Error('Select component: children can only contain Option components.');
+  }
+
+  const selectName = name || id;
+  const placeholder = '- Select option -';
+
+  const selectWrapperRef = useRef<HTMLDivElement>(null);
+  const selectButtonRef = useRef<HTMLButtonElement>(null);
+  const selectListRef = useRef<HTMLUListElement>(null);
+  const selectListHoverBackgroundRef = useRef<HTMLSpanElement>(null);
+  const filterSearchRef = useRef<HTMLInputElement>(null);
+
+  const [isOpen, setIsOpen] = useState(false);
+  const [selected, setSelected] = useState<string | null>(value);
+  const [selectListHeight, setSelectListHeight] = useState(0);
+  const [filterSearch, setFilterSearch] = useState('');
+
+  useEffect(() => {
+    setSelected(value);
+  }, [value]);
+
+  useEffect(() => {
+    const handleKeyDown = (e: KeyboardEvent) => {
```

Comment:
Move this function definition out of the useEffect for better readability 👍

Reply:
Too many dependencies, it will be a mess.